Adding DDP Design Note

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32158 Adding DDP Design Note**

Differential Revision: [D19405980](https://our.internmc.facebook.com/intern/diff/D19405980)

mrshenli committed:
Adding DDP Design Note

[ghstack-poisoned]

mrshenli commented:
Create permanent links for images

<img width="362" alt="ddp_code" src="https://user-images.githubusercontent.com/16999635/72313120-4e7c1c80-3658-11ea-9c6d-44336b2daeac.png">
<img width="715" alt="ddp_grad_sync" src="https://user-images.githubusercontent.com/16999635/72313121-4e7c1c80-3658-11ea-9e3a-ca75f697b9de.png">

mrshenli committed:
Update on "[WIP] Adding DDP Design Note"




[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit aa9e9648:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=aa9e96485c5109e5a7af24a9d7154826a14d7cf6).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4233452) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84850077) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84850077))
<details>
<summary>
<code>Jan 14 04:08:45 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 14 04:08:45 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 14 04:08:45 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 14 04:08:45 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 14 04:08:45 1 out of 1 hunk FAILED 
Jan 14 04:08:45 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 14 04:08:45 Traceback (most recent call last): 
Jan 14 04:08:45   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 14 04:08:45     patch_repo(args) 
Jan 14 04:08:45   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 14 04:08:45     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 14 04:08:45 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 14 04:08:45 + cleanup 
Jan 14 04:08:45 + retcode=1 
Jan 14 04:08:45 + set +x 
Jan 14 04:08:45 =================== sccache compilation log =================== 
Jan 14 04:08:45 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 14 04:08:45 Compile requests                 1 
Jan 14 04:08:45 Compile requests executed        0 
Jan 14 04:08:45 Cache hits                       0 
Jan 14 04:08:45 Cache misses                     0 
Jan 14 04:08:45 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

mrshenli committed:
Update on "[WIP] Adding DDP Design Note"




[ghstack-poisoned]

mrshenli commented:
Entire page view:

![page_view](https://user-images.githubusercontent.com/16999635/72402630-ba748880-371d-11ea-98e8-33f2ae731fa0.png)

pietern commented on docs/source/notes/ddp.rst:
```diff
@@ -0,0 +1,191 @@
+
+.. warning::
+  The implementation of :class:`torch.nn.parallel.DistributedDataParallel`
+  evolves over time. This design note is written based on the states as of v1.4.
```
/states/state/

pietern commented on docs/source/notes/ddp.rst:
```diff
@@ -0,0 +1,191 @@
+
+:class:`torch.nn.parallel.DistributedDataParallel` (DDP) replicates model,
+splits input data, and transparently performs distributed data parallel
+training. This page describes how it works and reveals implementation details.
```
Not entirely true, because it doesn't do input data splitting (done by the `DistributedSampler`). Also, it doesn't automatically replicate across machines (and I think we shouldn't promote usage of the built-in replication). I think saying "transparently performs distributed data parallel training of an arbitrary `nn.Module`" is enough here.

pietern commented on docs/source/notes/ddp.rst:
```diff
@@ -0,0 +1,191 @@
+  Hence, applications must create ``ProcessGroup`` instances before constructing
+  DDP.
+- **Construction**: The DDP constructor takes a reference to the local module,
+  and broadcasts ``state_dict()`` from rank0 process to all other processes in
```
from process with rank 0 to

pietern reviewed (approved):
Excellent!!

zhaojuanmao commented on docs/source/notes/ddp.rst:
```diff
@@ -0,0 +1,191 @@
+  the ``forward`` function for the ``nn.parallel.DistributedDataParallel``
+  module which call into C++ libraries. Its ``_sync_param`` function performs
+  intra-process parameter synchronization when one DDP process works on multiple
+  devices, and it also broadcast model buffers from rank0 to all other ranks.
```
broadcasts

zhaojuanmao reviewed (commented):
This is really great!!

mrshenli committed:
Update on "Adding DDP Design Note"


Differential Revision: [D19405980](https://our.internmc.facebook.com/intern/diff/D19405980)

[ghstack-poisoned]
